content/non-blocking-communication-pt1.rst

```diff
@@ -1,33 +1,307 @@
+
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be sent
+    to ``dest`` rank of ``comm`` using tag ``tag``. The ``request`` object
+    that is returned must be used to wait on the communication later.
+
+Other calls exist for other sending modes, including buffered,
+synchronous, and ready-mode sends.
+
+
+Non-blocking MPI receive call
+-----------------------------
 
+An |term-MPI_Irecv| creates a receive request and returns a request
+object.  The caller is responsible for not changing the buffer until
+after waiting upon the resulting request object.
 
-Topic introduction here
+Call signature:
 
-You really want to browse this page alongside the source of it, to see
-how this is implemented.  See the links at the to right of the page.
+.. code-block:: c
 
+    int MPI_Irecv(void* buf, int count, MPI_Datatype datatype,
+                  int source, int tag, MPI_Comm comm, MPI_Request *request)
 
 
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be
+    received from ``source`` rank of ``comm`` using tag ``tag``. The
+    ``request`` object that is returned must be used to wait on the
+    communication later.
+
+An |term-MPI_Irecv| can be used to match any kind of send, regardless of
+sending mode or blocking status.
+
+Waiting for non-blocking call completion
+----------------------------------------
+
+An |term-MPI_Wait| call waits for completion of the operation that
+created the request object passed to it. For a send, the semantics of
+the sending mode have been restored (not necessarily that the message
+has been received). For a receive, the buffer is now valid for use,
+however the send has not necessarily completed (though obviously has
+been initiated).
+
+Call signature:
+
+.. code-block:: c
+
+    int MPI_Wait(MPI_Request *request, MPI_Status *status)
+
+
+.. note::
+
+    ``request`` describes the operation to be waited upon. ``status``
+    returns the status of that operation. If the status is not needed,
+    pass ``MPI_STATUS_IGNORE``.
+
+It can be efficient to wait on any one, some, or all of a set of
+operations before returning. MPI provides |term-MPI_Waitany|,
+|term-MPI_Waitsome|, and |term-MPI_Waitall| for these use cases. For example,
+waiting for any request to complete may allow the caller to continue
+with related computation while waiting for other requests to complete.
+
+
+Testing for non-blocking call completion
+----------------------------------------
+
+An |term-MPI_Test| call returns immediately whether a corresponding
+|term-MPI_Wait| would return immediately. completes the operation that
+created the request object passed to it. For a send, the semantics of
+the sending mode have been restored (not necessarily that the message
+has been received). For a receive, the buffer is now valid for use,
+however the send has not necessarily completed (though obviously has
+been initiated).
```

Comment:
Is the copy-paste intended here?

Reply:
Good catch, I didn't clean up the copy-paste enough